### DISCUSSION TITLE
how to modify the request body in a middleware

### DISCUSSION THREAD

**Discussion by tobylaifun** [honojs/hono#3249] [General]
I am writing a bot for a platform called kook, so i am writing a webhook.
the request from the platform is compressed using zlib so i need to write a middleware to decompress the data, modify the request body and pass it to the app.
the middleware is like:
```typescript
createMiddleware(async (c, next) => {
    if (c.req.method != "POST") {
      await next();
      return;
    }
    const req = c.req.raw.clone();
   // some code
    let res="the decompressed text result, it's json text";
    let resReadableStream = new ReadableStream({
      start(controller) {
        controller.enqueue(new TextEncoder().encode(res));
        controller.close();
      },
    });
    c.req.raw = new Request(req, {
      body: resReadableStream,
      headers: { "Content-Type": "application/json" },
    });
    await next();
  });
```
but when i try to call `await c.req.json()` in `app.post("/foo",func)`, it throws `��N�0"... is not valid JSON 'x', "x�`. (i checked the `res` varible and found it was correct.)

After searching in the documentation, i didn't find the right way to modify the request body before `next()`(i only found the way to modify response after that). so i hope to find some clues here.

**Comment by tobylaifun** [comment]
Environment:
deno 1.43.6

**Reply by tobylaifun** [reply]
hono: 4.5.4

**Comment by tobylaifun** [comment]
it seems impossible so now i use `Context.set() / get()` instead

**Comment by rogeriopavan** [comment]
I dont now about alter all req, mas to add a value to a req i use:

`context.req["myVariable"] = "myvariable"`
